### PR TITLE
feat(learn): Add tests to Personal Library

### PR DESCRIPTION
**Pull request by SaintPeter**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

### Changes
* Adds all of the tests for all user stories, including my improved user stories (See #39630 and PR https://github.com/freeCodeCamp/boilerplate-project-library/pull/5)
~~* Removes InfoSec tests as per #39435~~

### Testing
I have tested this against my personal project. It will fail against the current boilerplate.

### CODE REVIEW

**Comment by SaintPeter** [comment]
I should have resolved the merge conflict.

**Review by naomi-lgbt — DISMISSED** [review]
This LGTM!

I tested locally against @SaintPeter's project, as the current example does fail. I am, however, in favour of adding the additional clarity to these test instructions. 

As such, prior to merging this I recommend ensuring an update on the example project.

**Comment by SaintPeter** [comment]
If you want to test against my solution:
https://personal-library.pearlygatesoftware.com/

**Review by Gimhan-minion — DISMISSED** [review]

**Comment by raisedadead** [comment]
This is blocked by #39024

**Comment by RandellDawson** [comment]
Fixed merge conflict, and PR #39024 has been merged, so removing `status: blocked` label.

**Comment by SaintPeter** [comment]
Fixed the merge conflict, added the instructions from the README file, as well as removing the user stories as per https://github.com/freeCodeCamp/freeCodeCamp/issues/39721.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:13]:
```suggestion

Build a full stack JavaScript app that is functionally similar to this: <a href='https://personal-library.freecodecamp.rocks/' target='_blank'>https://personal-library.freecodecamp.rocks/</a>. Working on this project will involve you writing your code using one of the following methods:

- Clone <a href='https://github.com/freeCodeCamp/boilerplate-personal-library'>this GitHub repo</a> and complete your project locally.
- Use <a href='https://repl.it/github/freeCodeCamp/boilerplate-personal-library' target='_blank'>our repl.it starter project</a> to complete your project.
- Use a site builder of your choice to complete the project. Be sure to incorporate all the files from our GitHub repo.

When you are done, submit a URL to a publicly visible working demo of your project in the `Solution Link` field and, optionally, submit a link to your projects source code in the `GitHub Link` field.

```

**Review by ShaunSHamilton — DISMISSED** [review]
The description for the projects has changed: 39721

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (moT01) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md]: Hey, sorry about that - I made one more comment regarding these to try and clarify the wording a little more. We may want to change it up a tiny bit yet.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:17]:
```suggestion
When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your projects source code in the `GitHub Link` field.
```

Comment (moT01) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:17]: Last one

**Review by moT01 — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:17]:
```suggestion
When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your project's source code in the `GitHub Link` field.
```

**Comment by Sky020** [comment]
Are the user stories going to be added to /learn in this PR?

**Comment by SaintPeter** [comment]
@Sky020 This PR adds the tests to the existing user stories on `/learn`.

**Comment by Sky020** [comment]
@SaintPeter Right. I was looking at the _most recent_ in the commit history.

Still, I was expecting to see the same _user stories_ which were on the `index.html` page. I am concerned many campers will not know where to begin, if all they see is the tests.

**Comment by SaintPeter** [comment]
@Sky020 While the most recent commit alters the instructions, the prior commits add tests to the existing user stories. As I understand it, the new standard is for the only instance of the user stories will be on the `/learn` page... Which they are.

It might make sense to include a specific instruction to the instructions section pointing them to the User Stories / Tests below the project submission buttons. Maybe a re-layout of the project page is needed, in light of this new standard?

**Comment by Sky020** [comment]
@SaintPeter I was thinking of [this comment](https://github.com/freeCodeCamp/freeCodeCamp/issues/39721#issuecomment-700703006):
> I think that, for now, we would be fine doing that and once we get everything moved over to learn, perhaps we can look into moving some user stories to the instructions area.

So, ignore my comments, for now. We should discuss it more officially, before doing anything. Thanks, for bearing with my scatter-brain.

**Comment by SaintPeter** [comment]
Figured out what was blocking the linter in the CI and updated the HTML style links to Markdown style links.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:82]:
```suggestion
  - text: I can <b>get</b> /api/books/{_id} to retrieve a single object of a book containing <code>title</code>, <code>_id</code>, & an array of <code>comments</code> (empty array if no comments present). If no book is found, return '<code>no book exists</code>'.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:102]:
```suggestion
  - text: I can <b>post</b> a <code>comment</code> to /api/books/{_id} to add a comment to a book and returned will be the books object similar to <b>get</b> /api/books/{_id}. If <code>comment</code> is not included in the request, return '<code>missing required field comment</code>'. If no book is found, return '<code>no book exists</code>`.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:130]:
```suggestion
  - text: I can <b>delete</b> /api/books/{_id} to delete a book from the collection. Returned will be '<code>delete successful</code>' if successful. If no book is found, return '<code>no book exists</code>'.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:123]:
```suggestion
         let failingComment = await $.post(url + '/5f665eb46e296f6b9b6a504d', { 'comment': 'Never Seen Comment' });
```
Missed the `5` here so doesn't cast to _id properly.

**Review by naomi-lgbt — DISMISSED** [review]
Some things I ran in to while working on an example project for this.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:43]:
```suggestion
  - text: You can send a <b>POST</b> request to <code>/api/books</code> with <code>title</code> as part of the form data to add a book.  The returned response will be an object with the <code>title</code> and a unique <code>_id</code> as keys.  If <code>title</code> is not included in the request, the returned response should be the string <code>missing required field title</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:58]:
```suggestion
  - text: You can send a <b>GET</b> request to <code>/api/books</code> and receive a JSON response representing all the books. 
 The JSON response will be an array of objects with each object (book) containing <code>title</code>, <code>_id</code>, and <code>commentcount</code> properties.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:82]:
```suggestion
  - text: You can send a <b>GET</b> request to <code>/api/books/{_id}</code> to retrieve a single object of a book containing the properties <code>title</code>, <code>_id</code>, and a <code>comments</code> array (empty array if no comments present). If no book is found, return the string <code>no book exists</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:102]:
```suggestion
  - text: You can send a <b>POST</b> request containing <code>comment</code> as the form body data to <code>/api/books/{_id}</code> to add a comment to a book. The returned response will be the books object similar to <b>GET</b> <code>/api/books/{_id}</code> request in an earlier test. If <code>comment</code> is not included in the request, return the string <code>missing required field comment</code>`. If no book is found, return the string <code>no book exists</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:130]:
```suggestion
  - text: You can send a <b>DELETE</b> request to <code>/api/books/{_id}</code> to delete a book from the collection. The returned response will be the string <code>delete successful</code> if successful. If no book is found, return the string <code>no book exists</code>.
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (RandellDawson) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:147]:
```suggestion
  - text: You can send a <b>DELETE</b> request to <code>/api/books</code> to delete all books in the database. The returned response will be the string <code>'complete delete successful</code> if successful.
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@SaintPeter @nhcarrigan I tested the changes against the changes in the example project PR (https://github.com/freeCodeCamp/demo-projects/pull/39) and everything looks fine.  I do have a few suggested changes. If you have a different way of wording what I have suggested, please feel free to let me know and I can update my suggestion on GitHub.

In addition to the suggested changes below, can you update the first test's `text` to the following to bring this project in line with the suggestions of issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721:
> You should provide your own project, not the example URL.

**Review by RandellDawson — DISMISSED** [review]

**Comment by RandellDawson** [comment]
I am good with these changes now.  @moT01 @Sky020 @nhcarrigan If you are good, then we can get this one merged.

**Comment by nhcarrigan** [comment]
I built an example project against this change and everything seemed good to me. I'm good with this.

**Review by RandellDawson — DISMISSED** [review]
@SaintPeter For the most part I am satisfied with the current changes.  Will still need to create a separate PR for this project and the boilerplate after merging both this PR and the [existing boilerplate PR](https://github.com/freeCodeCamp/boilerplate-project-library/pull/5) to add any further instructions that are present in the boilerplate comments and then remove those comments from the boilerplate to complete the requirements specified in issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721 .  I recommend looking at what was done in the Mocha/Chai challenges PR (#39976) by using `#` to reference the specific test the user should modify in the boilerplate.

For now, I am just have one suggested change for this PR.

**Review by moT01 — COMMENTED** [review]
Overall, these look real good @SaintPeter & @nhcarrigan 🎉 I am just going to leave my findings for all three PR's related to this project here.

I was going to mention what Randell brought up as well.

The boilerplate has some files in it that could be deleted - `unit-tests.js`, `.bashrc`, `.profile`, `.hyperdev-assets`. I'm quite sure these aren't needed - they don't have to be deleted before these can get merged - I will leave it up to you two if you want to do that. `unit-tests.js` also exists on the demo-project.

The demo project hangs for me when making any type of request to a bad id - e.g. https://personal-library.nicholascarriga.repl.co/api/books/bad_id - It does seem to work, however, if the "bad_id" is an id of a book that was deleted. So if I create a book, then delete it and use that id, it will give the correct `no book exists` response. I don't think this is a huge deal, but it might be nice to have the demo give out those responses. It also means that the tests are only checking ids that formerly existed or something? not sure. Thought I would point it out if you two want to take a look at all - perhaps you're not seeing the same thing. 

I would be fine merging all three of the PR's as-is if we want 🎉

**Comment by nhcarrigan** [comment]
> The demo project hangs for me when making any type of request to a bad id - e.g. https://personal-library.nicholascarriga.repl.co/api/books/bad_id - It does seem to work, however, if the "bad_id" is an id of a book that was deleted. So if I create a book, then delete it and use that id, it will give the correct no book exists response. I don't think this is a huge deal, but it might be nice to have the demo give out those responses. It also means that the tests are only checking ids that formerly existed or something? not sure. Thought I would point it out if you two want to take a look at all - perhaps you're not seeing the same thing.

This is likely a MongoDB "failed to cast to ObjectId" error - if the string isn't something that looks like an ObjectID, it gets mad. 
I can poke at this later the afternoon and update the demo project accordingly. 🙂

**Comment by nhcarrigan** [comment]
@moT01 I've added a check for valid ObjectID to the example project, which should resolve the error you encountered.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (moT01) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:28]: may want to change the 5 below this to 4 as well? - or you can also just turn all of them into 1 for an ordered list.

**Comment by SaintPeter** [comment]
I've made the requested changes.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md
Comment (nhcarrigan) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:28]:
```suggestion
1. Add your MongoDB connection string to `.env` without quotes as `DB`  
   Example: `DB=mongodb://admin:pass@1234.mlab.com:1234/fccpersonallib`
2. In your `.env` file set `NODE_ENV` to `test`, without quotes
3. You need to create all routes within `routes/api.js`
4. You will create all functional tests in `tests/2_functional-tests.js`
```

While markdown renders this correctly, it appears the linter doesn't like it.

**Review by moT01 — APPROVED** [review]
LGTM 🎉 Thanks for fixing this up @SaintPeter @nhcarrigan 🎉

**Pull request merged**
